### ISSUE TITLE
segfault during SSL negotiation with older NRPE 2.15 plugin and ssl_logging=-1

### ISSUE THREAD

**Issue by sebastic** [NagiosEnterprises/nrpe#91]
As reported by Adam Di Carlo in [Debian Bug #849417](https://bugs.debian.org/849417):
> Given a situation where a debian/stable (jessie) server is polling an NRPE node running the latest unstable NRPE server, with all debugging enabled (`ssl_logging=-1`), I am getting the following segfault, as reported in `/var/log/syslog`:
> 
> ```
> Dec 26 14:49:38 salsa nrpe[14736]: Connection from 192.168.1.5 port 59564
> Dec 26 14:49:38 salsa nrpe[14736]: Host address is in allowed_hosts
> Dec 26 14:49:38 salsa kernel: [176235.037105] nrpe[14736]: segfault at 50000335 ip 00007fd44f408496 sp 00007ffd5abfb418 error 4 in libc-2.24.so[7fd44f388000+195000]
> ```
> 
> However, if I rachet down the SSL debugging, e.g., `ssl_logging=0x03`, the segfault disappears. 

I can reproduce the issue when `nrpe` is started with SSL support (`/usr/sbin/nrpe -c /etc/nagios/nrpe.cfg -f`), but doesn't have an SSL certificate (`ssl_cert_file`) nor key (`ssl_privatekey_file`) configured in `/etc/nagios/nrpe.cfg`, for which syslog reports:

```
nrpe[15305]: SSL Certificate File: None
nrpe[15305]: SSL Private Key File: None
nrpe[15305]: SSL CA Certificate File: None
nrpe[15305]: SSL Cipher List: ALL:!MD5:@STRENGTH
nrpe[15305]: SSL Allow ADH: No
nrpe[15305]: SSL Client Certs: Don't Ask
nrpe[15305]: SSL Log Options: 0xffffffff
nrpe[15305]: SSL Version: TLSv1 And Above
nrpe[15305]: INFO: SSL/TLS initialized. All network traffic will be encrypted.
```

And also when an self-signed certificate is used (no CA certificate), for which syslog reports:
```
nrpe[15837]: Added command[check_total_procs]=/usr/lib/nagios/plugins/check_procs -w 150 -c 200
nrpe[15837]: SSL Certificate File: /etc/ssl/certs/ssl-cert-snakeoil.pem
nrpe[15837]: SSL Private Key File: /etc/ssl/private/ssl-cert-snakeoil.key
nrpe[15837]: SSL CA Certificate File: None
nrpe[15837]: SSL Cipher List: ALL:!MD5:@STRENGTH
nrpe[15837]: SSL Allow ADH: No
nrpe[15837]: SSL Client Certs: Don't Ask
nrpe[15837]: SSL Log Options: 0xffffffff
nrpe[15837]: SSL Version: TLSv1 And Above
nrpe[15837]: INFO: SSL/TLS initialized. All network traffic will be encrypted.
```

This is with OpenSSL 1.0.2j, testing with a newer version (as suggested in #63) is not possible since NRPE 3.x fails to build with OpenSSL 1.1.0.

**Comment by smutel** [comment]
Hello,

It seems that I have the same bug with nrpe 3.0.1.

```
> Feb 23 15:30:22 rcentol601 nrpe[24796]: Caught SIGTERM - shutting down...
> Feb 23 15:30:22 rcentol601 nrpe[24796]: Daemon shutdown
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL Certificate File: /opt/monitoring-agent/certs/monitoring-agent.pem
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL Private Key File: /opt/monitoring-agent/certs/monitoring-agent.key
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL CA Certificate File: /opt/monitoring-agent/certs/monitoring-agent-cert.pem
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL Cipher List: ADH
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL Allow ADH: Require
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL Client Certs: Require
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL Log Options: 0xffffffff
> Feb 23 15:30:22 rcentol601 nrpe[24860]: SSL Version: SSLv2 And Above
> Feb 23 15:30:22 rcentol601 nrpe[24861]: Starting up daemon
> Feb 23 15:30:22 rcentol601 nrpe[24861]: Server listening on 0.0.0.0 port 5666.
> Feb 23 15:30:22 rcentol601 nrpe[24861]: Server listening on :: port 5666.
> Feb 23 15:30:22 rcentol601 nrpe[24861]: Warning: Daemon is configured to accept command arguments from clients!
> Feb 23 15:30:22 rcentol601 nrpe[24861]: Listening for connections on port 5666
> Feb 23 15:30:22 rcentol601 nrpe[24861]: Allowing connections from: XX.XX.XX.XX
> Feb 23 15:30:36 rcentol601 kernel: nrpe[24866]: segfault at f7a5a0e8 ip 000000356ec47dec sp 00007ffe488b2b80 error 4 in libc-2.12.so[356ec00000+18a000]
> Feb 23 15:30:38 rcentol601 kernel: nrpe[24868]: segfault at f7a5a0e8 ip 000000356ec47dec sp 00007ffe488b2b80 error 4 in libc-2.12.so[356ec00000+18a000]
> Feb 23 15:32:42 rcentol601 kernel: nrpe[24874]: segfault at f7a5a0e8 ip 000000356ec47dec sp 00007ffe488b2b80 error 4 in libc-2.12.so[356ec00000+18a000]
```

So the SSL encryption is not working at all for me. The check_nrpe command returns "CHECK_NRPE: Error - Could not complete SSL handshake with XX.XX.XX.XX: 1".

Regards,
Samuel Mutel.

**Comment by jfrickson** [comment]
Looking into this

**Comment by jfrickson** [comment]
The fix for this is in branch `maint`. Fixed as a side-effect of fixing issue #75
